extensions/positron-python/src/client/positron/session.ts

```diff
@@ -538,35 +539,45 @@ export class PythonRuntimeSession implements positron.LanguageRuntimeSession, vs
         );
     }
 
-    async activateLsp() {
+    async activateLsp(): Promise<void> {
         // Start LSP for the foreground session only if its been previously stopped
         if (this._lsp?.state === LspState.stopped) {
-            this._queue.add(async () => {
-                const port = await this.adapterApi!.findAvailablePort([], 25);
-                if (this._kernel) {
-                    this._kernel.emitJupyterLog(`Starting Positron LSP server on port ${port}`);
-
-                    // Create the LSP comm before creating the LSP
-                    // client. We keep track of this initialisation in
-                    // case we need to restart, to avoid restarting in
-                    // the middle of init.
-                    this._lspStarting = this._kernel.startPositronLsp(`127.0.0.1:${port}`);
-
-                    await this._lspStarting;
-                    await this._lsp?.activate(port);
-                }
+            return this._queue.add(async () => {
+                await this.startLsp();
             });
+        } else {
+            return undefined;
         }
     }
 
-    deactivateLsp() {
+    async deactivateLsp(awaitStop: boolean): Promise<void> {
         if (this._lsp?.state === LspState.running) {
-            this._queue.add(async () => {
+            return this._queue.add(async () => {
                 if (this._kernel) {
                     this._kernel.emitJupyterLog(`Stopping Positron LSP server`);
                 }
-                await this._lsp?.deactivate(true);
+                await this._lsp?.deactivate(awaitStop);
             });
+        } else {
+            return undefined;
+        }
+    }
+
+    private async startLsp(): Promise<void> {
+        if (this._kernel) {
+            this._kernel.emitJupyterLog('Starting Positron LSP server');
+
+            // Create the LSP comm, which also starts the LSP server.
+            // We await the server selected port (the server selects the
+            // port since it is in charge of binding to it, which avoids
+            // race conditions). We also use this promise to avoid restarting
+            // in the middle of initialization.
+            this._lspStarting = this._kernel.startPositronLsp('127.0.0.1');
+            const port = await this._lspStarting;
+
+            this._kernel.emitJupyterLog(`Starting Positron LSP client on port ${port}`);
+
+            await this._lsp?.activate(port);
```

Comment:
i.e., ask the kernel to start the LSP Comm, pick a port and bind to it, then send that port back to us, and finally start up the LSP server. Then we get to connect to that port on our end and start up the LSP Client